BindException: Address already when using lambda_java runtime

### Expected Behavior

When updating micronaut and micronaut-aws in our project, we got a bind exception when trying to start the micronaut project locally.

I believe this issue was introduced with #2315. The `EmbeddedServerFactory.java` creates a new application context, which will cause a second `HttpServerEmbeddedServer` bean to be created, which tries to listen on the same port.

Full error

```
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
21:51:50.969 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Error starting HTTP server: Error instantiating bean of type  [io.micronaut.servlet.http.server.HttpServerEmbeddedServer]

Message: Address already in use
Path Taken:
new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, EmbeddedServer embeddedServer, ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
\---> new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, [EmbeddedServer embeddedServer], ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
      \---> new @j.i.Singleton i.m.s.h.s.HttpServerEmbeddedServer(ApplicationContext applicationContext, @Named("HttpServer") ApplicationContextProvider applicationContextProvider, ApplicationConfiguration applicationConfiguration, HttpServerConfiguration httpServerConfiguration, ApplicationEventPublisher<ServerShutdownEvent> serverShutdownEventPublisher, [HttpServer server])
io.micronaut.http.server.exceptions.HttpServerException: Error starting HTTP server: Error instantiating bean of type  [io.micronaut.servlet.http.server.HttpServerEmbeddedServer]

Message: Address already in use
Path Taken:
new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, EmbeddedServer embeddedServer, ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
\---> new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, [EmbeddedServer embeddedServer], ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
      \---> new @j.i.Singleton i.m.s.h.s.HttpServerEmbeddedServer(ApplicationContext applicationContext, @Named("HttpServer") ApplicationContextProvider applicationContextProvider, ApplicationConfiguration applicationConfiguration, HttpServerConfiguration httpServerConfiguration, ApplicationEventPublisher<ServerShutdownEvent> serverShutdownEventPublisher, [HttpServer server])
	at io.micronaut.servlet.http.server.AbstractServletServer.start(AbstractServletServer.java:88)
	at io.micronaut.servlet.http.server.AbstractServletServer.start(AbstractServletServer.java:33)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:81)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:347)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:322)
	at com.innovattic.ApplicationKt.main(Application.kt:8)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [io.micronaut.servlet.http.server.HttpServerEmbeddedServer]

Message: Address already in use
Path Taken:
new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, EmbeddedServer embeddedServer, ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
\---> new @j.i.Singleton i.m.s.h.s.DiscoveryServerStartedEventPublisher(ApplicationConfiguration applicationConfiguration, [EmbeddedServer embeddedServer], ApplicationEventPublisher<ServiceReadyEvent> applicationEventPublisher)
      \---> new @j.i.Singleton i.m.s.h.s.HttpServerEmbeddedServer(ApplicationContext applicationContext, @Named("HttpServer") ApplicationContextProvider applicationContextProvider, ApplicationConfiguration applicationConfiguration, HttpServerConfiguration httpServerConfiguration, ApplicationEventPublisher<ServerShutdownEvent> serverShutdownEventPublisher, [HttpServer server])
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2357)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3162)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:3012)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2774)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1786)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:210)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2137)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1367)
	at io.micronaut.servlet.http.server.$HttpServerEmbeddedServer$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2342)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3162)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:3012)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2774)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1786)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:210)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2137)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1367)
	at io.micronaut.servlet.http.server.$DiscoveryServerStartedEventPublisher$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2342)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3162)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:3012)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3625)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistrations(DefaultBeanContext.java:3580)
	at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3554)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1480)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:888)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:878)
	at io.micronaut.context.event.ApplicationEventPublisherFactory$2.lambda$$1(ApplicationEventPublisherFactory.java:215)
	at io.micronaut.core.util.SupplierUtil$1.get(SupplierUtil.java:47)
	at io.micronaut.context.event.ApplicationEventPublisherFactory$2.publishEvent(ApplicationEventPublisherFactory.java:226)
	at io.micronaut.context.DefaultBeanContext.publishEvent(DefaultBeanContext.java:1872)
	at io.micronaut.servlet.http.server.AbstractServletServer.start(AbstractServletServer.java:85)
	... 5 common frames omitted
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:567)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.netBind(ServerSocketChannelImpl.java:337)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:294)
	at java.base/sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:89)
	at jdk.httpserver/sun.net.httpserver.ServerImpl.<init>(ServerImpl.java:142)
	at jdk.httpserver/sun.net.httpserver.HttpServerImpl.<init>(HttpServerImpl.java:50)
	at jdk.httpserver/sun.net.httpserver.DefaultHttpServerProvider.createHttpServer(DefaultHttpServerProvider.java:35)
	at jdk.httpserver/com.sun.net.httpserver.HttpServer.create(HttpServer.java:150)
	at io.micronaut.servlet.http.server.HttpServerFactory.createHttpServer(HttpServerFactory.java:51)
	at io.micronaut.servlet.http.server.$HttpServerFactory$CreateHttpServer0$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2342)
	... 39 common frames omitted
```

### Actual Behaviour

Exception on startup.

### Steps To Reproduce

- create new empty micronaut project (i've tried 4.10.13)
- in gradle config, put `runtime("lambda_java")`
- now try to run the app with gradle: `./gradlew run`

### Environment Information

_No response_

### Example Application

_No response_

### Version

4.10.13

--- a/function-aws-api-proxy-test/src/main/java/io/micronaut/function/aws/proxy/test/EmbeddedServerFactory.java
+++ b/function-aws-api-proxy-test/src/main/java/io/micronaut/function/aws/proxy/test/EmbeddedServerFactory.java
@@ -21,11 +21,13 @@ import io.micronaut.context.ApplicationContextBuilder;
 import io.micronaut.context.ApplicationContextProvider;
 import io.micronaut.context.annotation.Factory;
 import io.micronaut.context.annotation.Requires;
+import io.micronaut.context.env.Environment;
 import io.micronaut.context.env.PropertySource;
 import io.micronaut.context.exceptions.ConfigurationException;
 import io.micronaut.core.annotation.Experimental;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.core.util.StringUtils;
+import io.micronaut.function.aws.MicronautLambdaContext;
 import io.micronaut.function.aws.proxy.payload2.APIGatewayV2HTTPEventFunction;
 import jakarta.inject.Named;
 import jakarta.inject.Singleton;
@@ -34,12 +36,14 @@ import jakarta.inject.Singleton;
 @Internal
 @Requires(property = "micronaut.function.aws.proxy.test.enabled", notEquals = StringUtils.FALSE)
 @Factory
+@Requires(notEnv = Environment.FUNCTION)
 class EmbeddedServerFactory {
 
     @Named("HttpServer")
     @Singleton
     ApplicationContextProvider httpServerApplicationContextProvider(ApplicationContext applicationContext) {
-        ApplicationContextBuilder builder = ApplicationContext.builder();
+        ApplicationContextBuilder builder = ApplicationContext.builder()
+            .environments(Environment.FUNCTION, MicronautLambdaContext.ENVIRONMENT_LAMBDA);
         for (PropertySource propertySource : applicationContext.getEnvironment().getPropertySources()) {
             builder = builder.propertySources(propertySource);
         }

--- a/function-aws-api-proxy-test/src/test/groovy/io/micronaut/function/aws/proxy/test/AwsApiProxyTestServerSpec.groovy
+++ b/function-aws-api-proxy-test/src/test/groovy/io/micronaut/function/aws/proxy/test/AwsApiProxyTestServerSpec.groovy
@@ -1,5 +1,12 @@
 package io.micronaut.function.aws.proxy.test
 
+import com.sun.net.httpserver.HttpServer
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.ApplicationContextProvider
+import io.micronaut.context.annotation.Property
+import io.micronaut.context.env.Environment
+import io.micronaut.function.aws.MicronautLambdaContext
+import io.micronaut.function.aws.proxy.payload2.APIGatewayV2HTTPEventFunction
 import io.micronaut.http.HttpRequest
 import io.micronaut.http.HttpResponse
 import io.micronaut.http.HttpStatus
@@ -12,19 +19,40 @@ import io.micronaut.http.annotation.QueryValue
 import io.micronaut.http.client.HttpClient
 import io.micronaut.http.client.annotation.Client
 import io.micronaut.http.uri.UriBuilder
+import io.micronaut.runtime.server.EmbeddedServer
 import io.micronaut.test.extensions.spock.annotation.MicronautTest
 import spock.lang.Issue
 import spock.lang.Specification
 
 import jakarta.inject.Inject
+import jakarta.inject.Named
 
 import java.nio.charset.StandardCharsets
 
 @MicronautTest
+@Property(name = 'micronaut.server.port', value = '18111')
 class AwsApiProxyTestServerSpec extends Specification {
     @Inject
     @Client('/')
     HttpClient client
+
+    @Inject
+    @Named('HttpServer')
+    ApplicationContextProvider applicationContextProvider
+
+    void 'nested Lambda context does not create an embedded server'() {
+        expect:
+        applicationContextProvider instanceof APIGatewayV2HTTPEventFunction
+
+        when:
+        ApplicationContext lambdaContext = applicationContextProvider.applicationContext
+
+        then:
+        lambdaContext.environment.activeNames.contains(Environment.FUNCTION)
+        lambdaContext.environment.activeNames.contains(MicronautLambdaContext.ENVIRONMENT_LAMBDA)
+        !lambdaContext.containsBean(EmbeddedServer)
+        !lambdaContext.containsBean(HttpServer)
+    }
 
     void 'test invoke function via server'() {
         when:

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -1,5 +1,5 @@
 [versions]
-micronaut = "5.1.3"
+micronaut = "5.1.5"
 micronaut-test = "5.0.1"
 
 bouncycastle = '1.70'
@@ -8,7 +8,7 @@ logback-json-classic = '0.1.5'
 
 micronaut-discovery = "5.0.0"
 micronaut-groovy = "5.0.0"
-micronaut-logging = "2.0.0"
+micronaut-logging = "2.0.1"
 micronaut-mongodb = "6.0.1"
 micronaut-reactor = "4.1.0"
 micronaut-serde = "3.0.0"

